Attachments: Tap to download looks huge on visionOS

rr-codes opened the pull request:
#### 5a50cf705fe831dbdf8333b654234d3256b114ef
<pre>
Attachments: Tap to download looks huge on visionOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=259094">https://bugs.webkit.org/show_bug.cgi?id=259094</a>
rdar://109599374

Reviewed by Tim Nguyen.

Adjust the size of the attachment element components to compensate for the scale difference of visionOS.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentIconSizeProperty):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(img#attachment-icon,):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(button#attachment-save-button):

Canonical link: <a href="https://commits.webkit.org/266038@main">https://commits.webkit.org/266038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/413c403d6132b30f45066997dd420272c7f86082

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12932 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13247 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14343 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12664 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14771 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13545 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14779 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10822 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11411 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18498 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11576 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14768 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12052 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9969 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11298 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3104 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15630 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -153,10 +171,17 @@ div#attachment-save-area {
+#if (defined(WTF_PLATFORM_VISION) && WTF_PLATFORM_VISION)
+    width: 28px;
+    height: 28px;
+    padding: 8px;
```
This padding looks redundant with line 184?

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -153,10 +171,17 @@ div#attachment-save-area {
+#endif
     background-color: -apple-system-tertiary-fill;
     border: 0;
+    padding: 8px;
```
Is the change to padding on iOS intentional?

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -105,10 +111,14 @@ div#attachment-information-block {
 }
 
 div#attachment-action {
+#if (defined(WTF_PLATFORM_VISION) && WTF_PLATFORM_VISION)
+    font-size: x-small;
+#else
+    font-size: small;
+#endif
```
Not new to your patch, but I feel like these should probably use dynamic type?

@squelart are the existing font sizes to spec?

squelart replied:
No, they were not explicitly specified, I went with sizes that visually matched the designer's mock-ups.

nt1m replied:
I opened a bug to convert the fonts to use -apple-system- ones fwiw

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -153,8 +177,14 @@ div#attachment-save-area {
+#if (defined(WTF_PLATFORM_VISION) && WTF_PLATFORM_VISION)
+    width: 28px;
+    height: 28px;
+    padding: 8px;
```
Please use `mask-size` on `button#attachment-save-button::before` to set the icon size like other platforms

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -207,6 +221,7 @@ void HTMLAttachmentElement::ensureWideLayoutShadowTree(ShadowRoot& root)
 
     m_containerElement = HTMLDivElement::create(document());
     m_containerElement->setIdAttribute(attachmentContainerIdentifier());
+    m_containerElement->setInlineStyleCustomProperty(attachmentIconSizeProperty(), makeString(attachmentIconSize, "px"));
```
```suggestion
    m_containerElement->setInlineStyleCustomProperty(attachmentIconSizeProperty(), makeString(attachmentIconSize, "px"_s));
```

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -170,6 +199,17 @@ button#attachment-save-button {
 }
 
 button#attachment-save-button::before {
+#if (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
+#if (defined(WTF_PLATFORM_VISION) && WTF_PLATFORM_VISION)
+    mask-size: 14px;
+#else
+    mask-size: 18px;
+#endif
+    background-color: -apple-system-blue;
+#else
+    mask-size: 14px;
+    background-color: -apple-system-secondary-label;
+#endif
```
```suggestion
#if (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
    background-color: -apple-system-blue;
#else
    background-color: -apple-system-secondary-label;
#endif
#if (defined(WTF_PLATFORM_IOS) && WTF_PLATFORM_IOS))
    mask-size: 18px;
#else
    mask-size: 14px;
#endif
```

Do you know if this works?

rr-codes replied:
I think it would, but I personally think what I have is more readable imo since it's more explicit about having vision specific things, since it's just a coincidence that macOS and visionOS have the same size.

squelart commented on LayoutTests/platform/ios-simulator/fast/attachment/cocoa/wide-attachment-rendering-expected.png:
Should this file really be added? (I thought rendering tests only produced .txt results, this file only being used to locally check the visual result.)

rr-codes replied:
how did this even happen lol

I will fix

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -59,11 +63,19 @@ img#attachment-icon,
+    justify-self: center;
+}
+
+div#attachment-placeholder,
```
I doubt this is the behavior we want for the placeholder? @squelart can you check?

squelart replied:
Yes we want that so that the placeholder has exactly the same size as icons.

nt1m replied:
The current patch makes it so the placeholder has 3/4th the size of the icons, not sure that's expected

squelart replied:
Yes, it's expected. The full square icon width is too wide (most icons are naturally 3/4, so we're aiming for that.)

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -24,14 +24,18 @@
-    --icon-size: 72px;
+#if (defined(WTF_PLATFORM_VISION) && WTF_PLATFORM_VISION)
+    --icon-margin: 8px;
+    --information-margin-end: 8px;
```
Does this mean we want the save button to be very close to the right edge?

rr-codes replied:
yes; the specification is 8px from the edge

squelart reviewed (approved):
LGTM.

squelart commented on LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
```diff
@@ -4,14 +4,14 @@ layer at (0,0) size 800x98
     RenderBody {BODY} at (8,8) size 784x82
       RenderAttachment {ATTACHMENT} at (1,1) size 266x80
-        RenderFlexibleBox {DIV} at (0,0) size 266x80 [bgcolor=#0000000D]
+        RenderFlexibleBox {DIV} at (0,0) size 266x80 [bgcolor=#0000000C]
```
Actually, I believe this should stay `#D`, as expected by online tests.
I also get `#C` when running tests locally, I'm not sure why... I should really investigate! 😅

pxlcoder replied:
@squelart the system color definitions probably changed between OS versions.

squelart commented on LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
```diff
@@ -4,48 +4,48 @@ layer at (0,0) size 785x678
+          text run at (0,26) width 47: "Blank: "
         RenderAttachment {ATTACHMENT} at (47,1) size 267x80
-          RenderFlexibleBox {DIV} at (0,0) size 266x80 [bgcolor=#0000000D]
+          RenderFlexibleBox {DIV} at (0,0) size 266x80 [bgcolor=#0000000C]
```
Keep the `#D`s as well.

squelart reviewed (changes requested):
Sorry, we need the rendering-expected colors, otherwise these tests will always fail online.

squelart reviewed (approved):
r+ mostly for the C++ parts, please wait for Tim's r+.

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -48,22 +52,30 @@ div#attachment-container {
 
 img#attachment-icon {
+    max-width: var(--icon-size);
+    max-height: var(--icon-size);
```
```suggestion
    height: var(--icon-size);
```

I would ensure the height is actually filled.

rr-codes replied:
but in the case of a wide and short icon, I wouldn't want the height to be filled, right?

nt1m replied:
we discussed this offline, and it seems like there is the possibility for wide icons, so this is fine

rr-codes committed:
Attachments: Tap to download looks huge on visionOS
https://bugs.webkit.org/show_bug.cgi?id=259094
rdar://109599374

Reviewed by Tim Nguyen.

Adjust the size of the attachment element components to compensate for the scale difference of visionOS.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentIconSizeProperty):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(img#attachment-icon,):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(button#attachment-save-button):

Canonical link: https://commits.webkit.org/266038@main
